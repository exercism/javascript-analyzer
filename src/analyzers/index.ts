import { BaseAnalyzer } from './base_analyzer'

import { Exercise } from '../exercise'
import { Solution } from '../solution'

import { get as getLogger } from '../utils/logger'

import path from 'path'

export class Analyzers {

  /**
   * Find the correct analyzer given an exercise
   *
   * @param exercise the exercise to lookup
   *
   * @throws if there is no such analyzer
   * @returns the constructor for the analyzer
   */
  public static find<T extends BaseAnalyzer>(exercise: Readonly<Exercise>): new (solution: Solution) => T  {
    const file = Analyzers.autoload(exercise)
    const key = Object.keys(file).find(key => file[key] instanceof Function)

    if (key === undefined) {
      throw new Error(`No Analyzer found in './${exercise.slug}`)
    }

    const analyzer = file[key]
    getLogger().log(`=> analyzer: ${analyzer.name}`)
    return analyzer
  }

  private static autoload(exercise: Readonly<Exercise>) {
    const modulePath = path.join(__dirname, exercise.slug, 'index') // explicit path (no extension)
    try {
      return require(modulePath)
    } catch(err) {
      const logger = getLogger()
      logger.error(`
Could not find the index.js analyzer in "${modulePath}"

Make sure that:
 - the slug "${exercise.slug}" is valid (hint: use dashes, not underscores)
 - there is actually an analyzer written for that exercise

Original error:

`.trimLeft())
      logger.fatal(JSON.stringify(err), -32)
    }
  }
}
