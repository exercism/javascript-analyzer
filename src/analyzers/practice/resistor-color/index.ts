import type { Input } from '@exercism/static-analysis'
import {
  AstParser,
  NoExportError,
  NoMethodError,
} from '@exercism/static-analysis'
import type { TSESTree } from '@typescript-eslint/typescript-estree'
import { IsolatedAnalyzerImpl } from '~src/analyzers/IsolatedAnalyzerImpl'
import { CommentType, factory } from '~src/comments/comment'
import {
  NO_METHOD,
  NO_NAMED_EXPORT,
  NO_PARAMETER,
  UNEXPECTED_PARAMETER,
} from '~src/comments/shared'
import type { WritableOutput } from '~src/interface'
import { ResistorColorSolution } from './ResistorColorSolution'

const TIP_EXPORT_INLINE = factory<'method.signature' | 'constant.signature'>`
Did you know that you can export functions, classes and constants directly
inline?
\`\`\`javascript
export ${'constant.signature'}

export ${'method.signature'}
\`\`\`
`('javascript.resistor-color.export_inline', CommentType.Informative)

const SIGNATURE_NOT_OPTIMAL = factory`
ðŸ“• If you look at the tests, the function \`colorCode\` only receives one
parameter. Nothing more and nothing less. Suggest that the student
removes the additional parameters from your function, as their value will
always be \`undefined\` or whatever default they've assigned.
`('javascript.resistor-color.signature_not_optimal', CommentType.Essential)

const USE_INDEX_OF = factory<'current'>`
ðŸ“• The analyzer expected \`indexOf\`, instead of  \`${'current'}\`.

ðŸ’¬ Replace \`${'current'}\` with a different built-in function, a function
which does exactly what is now explicitly programmed in: finding the index
of a given value in an \`Array\`.
`('javascript.resistor-color.use_index_of', CommentType.Essential)

const USE_ARRAY_COMPREHENSIONS = factory<'current'>`
ðŸ“• The analyzer expected \`indexOf\`, instead of  \`${'current'}\`.

ðŸ’¬ Replace \`${'current'}\` with a single function call that _directly_
finds the index of a given input value.
`('javascript.resistor-color.use_array_comprehensions', CommentType.Essential)

const USE_IMPLICIT_INCLUDES = factory<'current'>`
ðŸ“• Using \`includes\` iterates _twice_ over the array. Performance is not
an issue in this exercise, but the logic is unnecessary. **Please note**:
throwing an \`Error\`, when a color does not exist, is perfectly fine and
should not be discouraged.

ðŸ’¬ Remove the explicit existence check \`.includes\`. When the color
string is converted to a color code (number), there is already a special
value which is returned if the color is not present. Look for and use
that value instead.
`('javascript.resistor-color.use_implicit_includes', CommentType.Essential)

const DONT_NORMALISE_INPUTS = factory`
ðŸ“• It's always fine to be defensive about inputs, so that a program will
run correctly under more circumstances. However, arbitary input
normalisation is discouraged on Exercism. Later exercises will have
defined inputs that are denormalised where we'll expect sanitizing and
normalising the input. Keep it simple for now.

ðŸ’¬ Remove the call to \`.toLowerCase\`. The tests only provide the inputs
in lower case, and the colors should be defined in lower case. There is
no need to manually normalise the inputs.
`('javascript.resistor-color.dont_normalise_inputs', CommentType.Actionable)

type Program = TSESTree.Program

export class ResistorColorAnalyzer extends IsolatedAnalyzerImpl {
  protected async execute(input: Input, output: WritableOutput): Promise<void> {
    const [parsed] = await AstParser.ANALYZER.parse(input)

    // Firstly we want to check that the structure of this solution is correct
    // and that there is nothing structural stopping it from passing the tests
    const solution = this.checkStructure(parsed.program, parsed.source, output)

    // Now we want to ensure that the method signature is sane and that it has
    // valid arguments.
    this.checkSignature(solution, output)

    // There are a handful optimal solutions for gigasecond which needs no
    // comments and can just be approved. If we have it, then let's just
    // acknowledge it and get out of here.
    this.checkForOptimalSolutions(solution, output)

    // The solution might not be optimal but still be approvable. Check these
    // first and bail-out (with approval) if that's the case.
    this.checkForApprovableSolutions(solution, output)

    // Time to find sub-optimal code.
    this.checkForDisapprovables(solution, output)

    // The solution is automatically referred to the mentor if it reaches this
  }

  private checkStructure(
    program: Readonly<Program>,
    source: Readonly<string>,
    output: WritableOutput
  ): ResistorColorSolution | never {
    try {
      return new ResistorColorSolution(program, source)
    } catch (error: unknown) {
      if (error instanceof NoMethodError) {
        output.disapprove(NO_METHOD({ 'method.name': error.method }))
      }

      if (error instanceof NoExportError) {
        output.disapprove(NO_NAMED_EXPORT({ 'export.name': error.namedExport }))
      }

      throw error
    }
  }

  private checkSignature(
    { entry }: ResistorColorSolution,
    output: WritableOutput
  ): void | never {
    // If there is no parameter then this solution won't pass the tests.
    //
    if (!entry.hasAtLeastOneParameter) {
      output.disapprove(NO_PARAMETER({ 'function.name': entry.name }))
    }

    // If this is not a simple parameter, but something else such as a splat,
    // or a parameter with a default argument, bail out and refer to mentor.
    //
    if (!entry.hasSimpleParameter) {
      output.redirect(UNEXPECTED_PARAMETER({ type: entry.parameterType }))
    }

    // If there is more than one parameter, something fishy is going on. Collect
    // the comment for the student, to disapprove later on.
    //
    if (!entry.hasExactlyOneParameter) {
      output.add(SIGNATURE_NOT_OPTIMAL())
    }

    // TODO: do we want to check more here?

    //
    //
    if (output.hasCommentary) {
      output.disapprove()
    }
  }

  private checkForOptimalSolutions(
    solution: ResistorColorSolution,
    output: WritableOutput
  ): void | never {
    // The optional solution looks like this:
    //
    // const COLORS = [...]
    //
    // function colorCode(...) { ... }
    //
    // export function value([tens, ones]) {
    //   return colorCode(ones) + colorCode(tens) * 10
    // }
    //

    if (!solution.isOptimal()) {
      // continue analyzing
      this.logger.log('~> solution is not optimal')
      return
    }

    this.checkForTips(solution, output)
    output.approve()
  }

  private checkForApprovableSolutions(
    solution: ResistorColorSolution,
    output: WritableOutput
  ): void | never {
    //
  }

  private checkForDisapprovables(
    solution: ResistorColorSolution,
    output: WritableOutput
  ): void | never {
    const numberOfComments = output.comments.length

    if (solution.entry.hasFindIndex) {
      output.add(USE_INDEX_OF({ current: '.findIndex(...)' }))
    } else if (solution.entry.hasForEach) {
      output.add(USE_ARRAY_COMPREHENSIONS({ current: '.forEach(...)' }))
    } else if (solution.entry.hasFor) {
      output.add(USE_ARRAY_COMPREHENSIONS({ current: 'for(...) { }' }))
    }

    if (solution.entry.hasIncludes) {
      output.add(USE_IMPLICIT_INCLUDES())
    }

    if (solution.entry.hasToLowerCase) {
      output.add(DONT_NORMALISE_INPUTS())
    }

    if (numberOfComments < output.commentCount) {
      // output.comments.unshift(BETA_COMMENTARY_PREFIX())
      output.disapprove()
    }

    //
  }

  private checkForTips(
    solution: ResistorColorSolution,
    output: WritableOutput
  ): void | never {
    if (!solution.hasInlineExports) {
      if (!output.hasCommentary) {
        // output.add(BETA_COMMENTARY_PREFIX())
      }

      // export { gigasecond }
      output.add(
        TIP_EXPORT_INLINE({
          'method.signature': solution.entry.signature,
          'constant.signature': solution.constant.signature,
        })
      )
    }

    // TODO optimize param
    // TODO use helper method
  }
}
