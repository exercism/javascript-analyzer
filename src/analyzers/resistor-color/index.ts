import { AST_NODE_TYPES } from "@typescript-eslint/typescript-estree";
import { Program } from "@typescript-eslint/typescript-estree/dist/ts-estree/ts-estree";

import { isReturnBlockStatement } from "~src/analyzers//utils/is_return_block_statement";
import { isReturnStatementWithValue } from "~src/analyzers//utils/is_return_statement_with_value";
import { AnalyzerImpl } from "~src/analyzers/AnalyzerImpl";
import { extractExport } from "~src/analyzers/utils/extract_export";
import { extractMainMethod, MainMethod } from "~src/analyzers/utils/extract_main_method";
import { parameterName } from '~src/analyzers/utils/extract_parameter';
import { findTopLevelConstants } from "~src/analyzers/utils/find_top_level_constants";
import { isCallExpression } from "~src/analyzers/utils/is_call_expression";
import { isIdentifier } from "~src/analyzers/utils/is_identifier";
import { annotateType } from "~src/analyzers/utils/type_annotations";
import { factory } from "~src/comments/comment";
import { NO_METHOD, NO_NAMED_EXPORT, NO_PARAMETER, UNEXPECTED_SPLAT_ARGS, PARSE_ERROR } from "~src/comments/shared";
import { AstParser, ParsedSource } from "~src/parsers/AstParser";
import { NoSourceError } from "~src/errors/NoSourceError";
import { ParserError } from "~src/errors/ParserError";

const TIP_EXPORT_INLINE = factory`
Did you know that you can export functions, classes and constants directly
inline?
`('javascript.resistor-color.export_inline')

const Parser: AstParser = new AstParser(undefined, 1)

const NOT_FOUND = {} as const

export class ResistorColorAnalyzer extends AnalyzerImpl {

  private program!: Program
  private source!: string

  private _mainMethod!: ReturnType<typeof extractMainMethod>
  private _mainConstant!: ReturnType<typeof findTopLevelConstants>[0] | typeof NOT_FOUND
  private _mainExports!: { function: ReturnType<typeof extractExport>; constant: ReturnType<typeof extractExport> }

  private get mainMethod(): ReturnType<typeof extractMainMethod> {
    if (!this._mainMethod) {
      this._mainMethod = extractMainMethod(this.program, 'colorCode')
    }
    return this._mainMethod
  }

  private get mainExports(): { function: ReturnType<typeof extractExport>; constant: ReturnType<typeof extractExport> } {
    if (!this._mainExports) {
      this._mainExports = {
        function: extractExport(this.program, 'colorCode'),
        constant: extractExport(this.program, 'COLORS')
      }
    }
    return this._mainExports
  }

  private get mainConstant(): ReturnType<typeof findTopLevelConstants>[0] | typeof NOT_FOUND {
    if (!this._mainConstant) {
      this._mainConstant = findTopLevelConstants(this.program, ['let', 'const', 'var']).find(
        ({ declarations }): boolean => !!declarations.find((declaration): boolean => {
          return isIdentifier(declaration.id, 'COLORS')
        })
      ) || NOT_FOUND
    }

    return this._mainConstant
  }

  public async execute(input: Input): Promise<void> {
    const [parsed] = await this.parse(input)

    this.program = parsed.program
    this.source = parsed.source

    // Firstly we want to check that the structure of this solution is correct
    // and that there is nothing structural stopping it from passing the tests
    this.checkStructure()

    // Now we want to ensure that the method signature is sane and that it has
    // valid arguments
    this.checkSignature()

    // There are a handful optimal solutions for resistor-color which needs no
    // comments and can just be approved. If we have it, then let's just
    // acknowledge it and get out of here.
    this.checkForOptimalSolutions()

    // The solution is automatically referred to the mentor if it reaches this
  }

  private async parse(input: Input): never | Promise<ParsedSource[]> {
    try {
      return await Parser.parse(input)
    } catch (err) {
      if (err instanceof NoSourceError) {
        this.logger.error(`=> [NoSourceError] ${err.message}`)
        this.redirect()
      }

      if (err instanceof ParserError) {
        this.logger.error(`=> [ParserError] ${err.message}`)
        const { message, ...details } = err.original
        this.disapprove(PARSE_ERROR({ error: message, details: JSON.stringify(details) }))
      }

      throw err
    }
  }

  private checkStructure(): void | never {
    const method = this.mainMethod
    const { function: [functionDeclaration,], constant: [constantDeclaration, ] } = this.mainExports

    // First we check that there is a two-fer function and that this function
    // is exported.
    if (!method) {
      this.comment(NO_METHOD({ 'method.name': 'colorCode' }))
    }

    if (!functionDeclaration) {
      this.comment(NO_NAMED_EXPORT({ 'export.name': 'colorCode' }))
    }

    if (!constantDeclaration) {
      this.comment(NO_NAMED_EXPORT({ 'export.name': 'COLORS' }))
    }

    if (this.hasCommentary) {
      this.disapprove()
    }
  }

  private checkSignature(): void | never {
    const method: MainMethod = this.mainMethod!

    // If there is no parameter
    // then this solution won't pass the tests.
    if (method.params.length === 0) {
      this.disapprove(NO_PARAMETER({ 'function.name': method.id!.name }))
    }

    const firstParameter = method.params[0]

    // If they provide a splat, the tests can pass but we should suggest they
    // use a real parameter.
    if (firstParameter.type === AST_NODE_TYPES.RestElement) {
      const splatArgName = parameterName(firstParameter)
      const splatArgType = annotateType(firstParameter.typeAnnotation)

      this.disapprove(UNEXPECTED_SPLAT_ARGS({
        'splat-arg.name': splatArgName,
        'parameter.type': splatArgType
      }))
    }
  }

  private checkForOptimalSolutions(): void | never {
    // The optional solution looks like this:
    //
    // export const COLORS = ['...', '...']
    // export function colorCode(color) {
    //   return COLORS.indexOf(color)
    // }
    //
    // The COLORS constant must be an Array and there must be a single call to
    // indexOf. Additionally, the function can not have a default argument, or
    // any other syntax, such as conditionals, re-assignment, and so forth.
    //
    // Other solutions might be approved but this
    // is the only one that we would approve without comment.
    //

    if (
    // !this.isArgumentOptimal()
      !this.isOneCallSolution()
      // || !this.isUsingArrayColors()
      // || !this.isUsingIndexOf()
    ) {
      // continue analyzing
      this.logger.log('~> Solution is not optimal')
      return
    }

    this.checkForTips()
    this.approve()
  }

  private checkForTips(): void | never {
    if (!this.hasInlineExport()) {
      this.comment(TIP_EXPORT_INLINE())
    }
  }

  private isOneCallSolution(): boolean | never {
    // Maximum body count may be 3 (3 - 1 + 1)
    //
    // 1: export function colorCode(color) {
    // 2:  return ...
    // 3: }
    //
    // but can also be less 1 (1 - 1 + 1)
    //
    // 1: export const colorCode = (color) => ...
    //
    const method: MainMethod = this.mainMethod!
    const body = method.body!

    // This trick actually looks to the inner exppresion instead of the entire
    // function in order to allow for comments inside the body.
    const { loc: { start: { line: lineStart }, end: { line: lineEnd } } } =
      isReturnBlockStatement(body)
        ? body.body[0]
        : method

    this.logger.log(`=> Body consists of ${lineEnd - lineStart + 1} lines and is a ${method.type}.`)

    if ((lineEnd - lineStart + 1) > 3) {
      return false
    }

    // There can only be a single call to COLORS.indexOf and there may not be
    // an other tokens, whatsoever. The following checks are only testing the
    // optimal variations.

    // function colorCode(color) { return COLORS.indexOf(color) }
    //
    if (method.type === AST_NODE_TYPES.FunctionDeclaration) {
      this.logger.log(`=> The Function Declaration has ${method.params.length} argument(s)`)

      if (method.params.length !== 1) {
        return false
      }

      const functionParameter = method.params[0]
      const callExpression = method.body
        && isReturnBlockStatement(method.body)
        && isReturnStatementWithValue(method.body.body[0])
        && isCallExpression(method.body.body[0].argument, 'COLORS', 'indexOf')
        && method.body.body[0].argument

      return !!(
        isIdentifier(functionParameter)
        && callExpression
        && isIdentifier(callExpression.arguments[0], functionParameter.name)
      )
    }

    // const colorCode = (color) => COLORS.indexOf(color)
    // const colorCode = (color) => { return COLORS.indexOf(color) }
    //
    if (method.type === AST_NODE_TYPES.ArrowFunctionExpression) {
      this.logger.log(`=> The Arrow Function Expression has ${method.params.length} argument(s)`)

      if (method.params.length !== 1) {
        return false
      }

      const functionParameter = method.params[0]

      // First test/option is the implicit return
      // Second test/option is the explicit return
      //
      const callExpression = (isCallExpression(body, 'COLORS', 'indexOf') && body)
        || (
          isReturnBlockStatement(body)
          && isReturnStatementWithValue(body.body[0])
          && isCallExpression(body.body[0].argument, 'COLORS', 'indexOf')
          && body.body[0].argument
        )

      return isIdentifier(functionParameter)
        && callExpression
        && isIdentifier(callExpression.arguments[0], functionParameter.name)
    }

    // const colorCode = function (color) { return COLORS.indexOf(color) }
    //
    if (method.type === AST_NODE_TYPES.FunctionExpression) {
      this.logger.log(`=> The Function Expression has ${method.params.length} argument(s)`)

      if (method.params.length > 1) {
        return false
      }

      const functionParameter = method.params[0]
      const callExpression = body
        && isReturnBlockStatement(body)
        && isReturnStatementWithValue(body.body[0])
        && isCallExpression(body.body[0].argument, 'COLORS', 'indexOf')
        && body.body[0].argument

      return isIdentifier(functionParameter)
        && callExpression
        && isIdentifier(callExpression.arguments[0], functionParameter.name)
    }

    // Should _never_ happen
    this.logger.log(`=> The body failed all the stuctural tests. It's a ${method!.type}.`)
    // Bail out
    return this.redirect()
  }

  private hasInlineExport(): boolean  {
    // Additionally make sure the export is inline by checking if it doesn't
    // have any specifiers:
    //
    // export function value
    // => no specififers
    //
    // export { value }
    // => yes specififers
    //
    return this.mainExports.function[0]!.specifiers
      && this.mainExports.function[0]!.specifiers.length === 0
      && this.mainExports.constant[0]!.specifiers
      && this.mainExports.constant[0]!.specifiers.length === 0
  }
}

