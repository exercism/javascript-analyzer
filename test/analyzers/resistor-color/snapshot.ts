import { ResistorColorAnalyzer } from '~src/analyzers/resistor-color'
import { makeTestGenerator } from '~test/helpers/snapshot'

const snapshotTestsGenerator = makeTestGenerator(
  'resistor-color', () => new ResistorColorAnalyzer()
)

describe('When running analysis on two-fer fixtures', () => {
  snapshotTestsGenerator('approve', [
    0, 1, 100, 101, 102, 105, 106, 107, 110, 113,
    115, 117, 12, 122, 123, 127, 13, 131, 132, 134,
    135, 137, 138, 14, 142, 149, 150, 151, 158, 16,
    161, 163, 164, 165, 168, 17, 171, 172, 173, 174,
    175, 179, 18, 180, 182, 184, 186, 187, 189, 191,
    192, 196, 197, 199, 2, 20, 206, 209, 211, 212
  ])
  snapshotTestsGenerator('disapprove', [])
  snapshotTestsGenerator('refer_to_mentor', [])
})
