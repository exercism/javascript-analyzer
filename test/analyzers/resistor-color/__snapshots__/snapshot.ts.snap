// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/0's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/1's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/2's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/12's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/13's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/14's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/16's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/17's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/18's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/20's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/100's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/101's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/102's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/105's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/106's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/107's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/110's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/113's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/115's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/117's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/122's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/123's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/127's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/131's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/132's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/134's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/135's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/137's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/138's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/142's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/149's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/150's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/151's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/158's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/161's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/163's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/164's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/165's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/168's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/171's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/172's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/173's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/174's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/175's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/179's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/180's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/182's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/184's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/186's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/187's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/189's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/191's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/192's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/196's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/197's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/199's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/206's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/209's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/211's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;

exports[`When running analysis on two-fer fixtures and expecting it to approve as optimal matches resistor-color/212's output: output 1`] = `
AnalyzerOutput {
  "comments": Array [],
  "status": "approve_as_optimal",
}
`;
